### ISSUE TITLE
Controller server + local costmap setup issues

### ISSUE THREAD

**Issue by lucasl84** [ros-navigation/navigation2#4532]
Hi everyone,

I've a few questions to which I can't find a reasonable answer about the controller server and the planner server. I've been looking at the turtlebot example just as a starting point but as my launch files are considerable different from the example, most of what is done there is not directly applicable to my setup.

I'm on ROS2 Galactic running on Ubuntu 20.04 (though I think my concerns are independent of concrete versions of both ROS and the OS).

When opening the ` nav2_bringup/bringup/params/nav2_params.yaml` file I see several blocks, among these are the controller server and local costmap configurations:

```yaml
controller_server:
  ros__parameters:
    use_sim_time: True
    controller_frequency: 20.0
    min_x_velocity_threshold: 0.001
    min_y_velocity_threshold: 0.5
    min_theta_velocity_threshold: 0.001
    failure_tolerance: 0.3
    progress_checker_plugin: "progress_checker"
    goal_checker_plugins: ["general_goal_checker"] # "precise_goal_checker"
    controller_plugins: ["FollowPath"]

    # Progress checker parameters
    progress_checker:
      plugin: "nav2_controller::SimpleProgressChecker"
      required_movement_radius: 0.5
      movement_time_allowance: 10.0
    # Goal checker parameters
    #precise_goal_checker:
    #  plugin: "nav2_controller::SimpleGoalChecker"
    #  xy_goal_tolerance: 0.25
    #  yaw_goal_tolerance: 0.25
    #  stateful: True
    general_goal_checker:
      stateful: True
      plugin: "nav2_controller::SimpleGoalChecker"
      xy_goal_tolerance: 0.25
      yaw_goal_tolerance: 0.25
    # DWB parameters
    FollowPath:
      plugin: "dwb_core::DWBLocalPlanner"
      debug_trajectory_details: True
      min_vel_x: 0.0
      min_vel_y: 0.0
      max_vel_x: 0.26
      max_vel_y: 0.0
      max_vel_theta: 1.0
      min_speed_xy: 0.0
      max_speed_xy: 0.26
      min_speed_theta: 0.0
      # Add high threshold velocity for turtlebot 3 issue.
      # https://github.com/ROBOTIS-GIT/turtlebot3_simulations/issues/75
      acc_lim_x: 2.5
      acc_lim_y: 0.0
      acc_lim_theta: 3.2
      decel_lim_x: -2.5
      decel_lim_y: 0.0
      decel_lim_theta: -3.2
      vx_samples: 20
      vy_samples: 5
      vtheta_samples: 20
      sim_time: 1.7
      linear_granularity: 0.05
      angular_granularity: 0.025
      transform_tolerance: 0.2
      xy_goal_tolerance: 0.25
      trans_stopped_velocity: 0.25
      short_circuit_trajectory_evaluation: True
      stateful: True
      critics: ["RotateToGoal", "Oscillation", "BaseObstacle", "GoalAlign", "PathAlign", "PathDist", "GoalDist"]
      BaseObstacle.scale: 0.02
      PathAlign.scale: 32.0
      PathAlign.forward_point_distance: 0.1
      GoalAlign.scale: 24.0
      GoalAlign.forward_point_distance: 0.1
      PathDist.scale: 32.0
      GoalDist.scale: 24.0
      RotateToGoal.scale: 32.0
      RotateToGoal.slowing_factor: 5.0
      RotateToGoal.lookahead_time: -1.0
      
local_costmap:
  local_costmap:
    ros__parameters:
      update_frequency: 5.0
      publish_frequency: 2.0
      global_frame: odom
      robot_base_frame: base_link
      use_sim_time: True
      rolling_window: true
      width: 3
      height: 3
      resolution: 0.05
      robot_radius: 0.22
      plugins: ["voxel_layer", "inflation_layer"]
      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        cost_scaling_factor: 3.0
        inflation_radius: 0.55
      voxel_layer:
        plugin: "nav2_costmap_2d::VoxelLayer"
        enabled: True
        publish_voxel_map: True
        origin_z: 0.0
        z_resolution: 0.05
        z_voxels: 16
        max_obstacle_height: 2.0
        mark_threshold: 0
        observation_sources: scan
        scan:
          topic: /scan
          max_obstacle_height: 2.0
          clearing: True
          marking: True
          data_type: "LaserScan"
          raytrace_max_range: 3.0
          raytrace_min_range: 0.0
          obstacle_max_range: 2.5
          obstacle_min_range: 0.0
      static_layer:
        map_subscribe_transient_local: True
      always_send_full_costmap: True
      
  local_costmap_client:
    ros__parameters:
      use_sim_time: True
      
  local_costmap_rclcpp_node:
    ros__parameters:
      use_sim_time: True
```

When inspecting the ` nav2_bringup/bringup/launch/localization_launch.py` file I see this:

```python
def generate_launch_description():
    # Get the launch directory
    bringup_dir = get_package_share_directory('nav2_bringup')

    namespace = LaunchConfiguration('namespace')
    use_sim_time = LaunchConfiguration('use_sim_time')
    autostart = LaunchConfiguration('autostart')
    params_file = LaunchConfiguration('params_file')

    lifecycle_nodes = ['controller_server',
                       'planner_server',
                       'recoveries_server',
                       'bt_navigator',
                       'waypoint_follower']

    # Map fully qualified names to relative ones so the node's namespace can be prepended.
    # In case of the transforms (tf), currently, there doesn't seem to be a better alternative
    # https://github.com/ros/geometry2/issues/32
    # https://github.com/ros/robot_state_publisher/pull/30
    # TODO(orduno) Substitute with `PushNodeRemapping`
    #              https://github.com/ros2/launch_ros/issues/56
    remappings = [('/tf', 'tf'),
                  ('/tf_static', 'tf_static')]

    # Create our own temporary YAML files that include substitutions
    param_substitutions = {
        'use_sim_time': use_sim_time,
        'autostart': autostart}

    configured_params = RewrittenYaml(
            source_file=params_file,
            root_key=namespace,
            param_rewrites=param_substitutions,
            convert_types=True)

    return LaunchDescription([
        # Set env var to print messages to stdout immediately
        SetEnvironmentVariable('RCUTILS_LOGGING_BUFFERED_STREAM', '1'),

        DeclareLaunchArgument(
            'namespace', default_value='',
            description='Top-level namespace'),

        DeclareLaunchArgument(
            'use_sim_time', default_value='false',
            description='Use simulation (Gazebo) clock if true'),

        DeclareLaunchArgument(
            'autostart', default_value='true',
            description='Automatically startup the nav2 stack'),

        DeclareLaunchArgument(
            'params_file',
            default_value=os.path.join(bringup_dir, 'params', 'nav2_params.yaml'),
            description='Full path to the ROS2 parameters file to use'),

        Node(
            package='nav2_controller',
            executable='controller_server',
            output='screen',
            parameters=[configured_params],
            remappings=remappings),

        Node(
            package='nav2_planner',
            executable='planner_server',
            name='planner_server',
            output='screen',
            parameters=[configured_params],
            remappings=remappings),

        Node(
            package='nav2_recoveries',
            executable='recoveries_server',
            name='recoveries_server',
            output='screen',
            parameters=[configured_params],
            remappings=remappings),

        Node(
            package='nav2_bt_navigator',
            executable='bt_navigator',
            name='bt_navigator',
            output='screen',
            parameters=[configured_params],
            remappings=remappings),

        Node(
            package='nav2_waypoint_follower',
            executable='waypoint_follower',
            name='waypoint_follower',
            output='screen',
            parameters=[configured_params],
            remappings=remappings),

        Node(
            package='nav2_lifecycle_manager',
            executable='lifecycle_manager',
            name='lifecycle_manager_navigation',
            output='screen',
            parameters=[{'use_sim_time': use_sim_time},
                        {'autostart': autostart},
                        {'node_names': lifecycle_nodes}]),

    ])
```

So my first question is this: How does the configuration of the local costmap is actually passed to the controller server which in turn spawns the local costmap? I'm asking this because I'm not using this code to launch the navigation stack but a different one where I manually read yaml configuration files and adjust any required parameters accordingly just before spawning a relevant node.

This has worked without issues for me when using SLAM toolbox, AMCL and the map server. For example, in my custom setup when constructing the AMCL node I do something like this (showing AMCL but this worked for any other ROS2 nodes that I've instantiated):

```python
def get_amcl(namespace=None):
    package_dir_name = 'MY_PACKAGE'
    yaml = read_yaml(package_dir_name, 'amcl.yaml')
    ros_parameters = yaml['amcl']['ros__parameters']

    if namespace:
        ros_parameters.update({
            'scan_topic': f"/{namespace}/{ros_parameters['scan_topic']}",
            'map_topic': f"/map_server",
            'base_frame_id': f'{namespace}/base_link',
            'global_frame_id': f'map',
            'odom_frame_id': f'{namespace}/odom',
        })

    options = {
        'package': 'nav2_amcl',
        'executable': 'amcl',
        'name': 'nav2_amcl',
        'output': 'screen',
        'parameters': [ros_parameters],
        'remappings': [
            ('/tf', f'/{namespace}/tf'),
            ('/tf_static', f'/{namespace}/tf_static'),
        ],
    }

    if namespace:
        options['namespace'] = namespace

    return Node(**options)
```

So I'm expecting to be able to configure the controller server in a similar way but how to pass BOTH configurations (controller server config and local costmap config) to the controller server node is not clear to me.

To be concrete: When creating a controller server how do I pass the configuration to the "hosted" local costmap that is automatically spawning? Analogously I have the same issue for the planner server and the global costmap.

According to this: [controller server](https://docs.nav2.org/configuration/packages/configuring-controller-server.html), the documentation does mention that indeed the controller server hosts a local costmap but it does not mention at all the mechanism by which the local costmap configuration is fed to it, the example at the bottom only shows the controller server config but not how its hosted local costmap should be configured, nor how these settings are received is explained nor mentioned.

What I'm experiencing at the moment are not errors, but misconfigurations as I need to modify the local costmap parameters (the frames) and somehow feed them to the hosted local costmap in the controller server.

Then another thing that I also found very confusing is the fact that both the local costmap and the global costmap configuration blocks start like this:

```yaml
local_costmap:
  local_costmap:
    ros__parameters:
      ...
      
global_costmap:
  global_costmap:
    ros__parameters:
      ...
```
Why are the `local_costmap` and `global_costmap` keys nested/repeated twice? What does that mean? All the rest of the configuration
blocks do not have these nestings. There's no comment on the YAML files as to why this is case, nor I could find information related to this. In which cases I'm expected to have those repetitions? What do they solve?

Thanks for everyone's time,
Lucas.

**Comment by SteveMacenski** [comment]
>  How does the configuration of the local costmap is actually passed to the controller server which in turn spawns the local costmap? 

`params_file = LaunchConfiguration('params_file')` gets the commandline argument value for `params_file`, which should be a path to your params file. Below declares that launch configuration and provides a default value if not otherwise set by the user. If you want to change and use this launch file you could append `params_file:=/path/to/your/file.yaml` and that default value is replaced and sent off to all the servers. 

```
        DeclareLaunchArgument(
            'params_file',
            default_value=os.path.join(bringup_dir, 'params', 'nav2_params.yaml'),
            description='Full path to the ROS2 parameters file to use'),
```

> but it does not mention at all the mechanism by which the local costmap configuration is fed to it

The local/global costmaps are hosted by their respective servers and they create their own `Node` objects independently from the servers to process data so that they act separately but can access each other's information without network latency. Thus, the costmaps read their own parameters themselves without any passing along. Send the controller server process a file (or 2 files is also possible in ROS 2 parameter fields) with both the local costmap and the controller server and you're good to go. If you've shown your tool works for AMCL/Map Server, then this should be no different :-) 

> Why are the local_costmap and global_costmap keys nested/repeated twice? What does that mean? All the rest of the configuration

The nodes are namespaced by their own name for subtle but important reasons related to the topic handling and making sure we don't have conflicts between services/topics that the costmaps expose when multiple costmaps are up in a system (like, for instance the local and global costmap are just 2 instances of the same thing just configured differently -- or adding in a 3rd or 4th for new servers, etc). 

Hope that answers your questions

**Comment by lucasl84** [comment]
Hi Steve,

First wanted to thank you for your fast reply. Unfortunately I'm still not getting this. I'm going to explain step by step how is it that I'm spawning the controller node because maybe I wasn't clear enough. Let's assume for a second that we do NOT have available any of the helpers provided by ROS (no DeclareLaunchArgument, LaunchDescription, LaunchConfiguration, etc) and just a `read_yaml` function that reads a YAML file and leaves its content in a dictionary:

```python
import yaml

def read_yaml(filename):
    with open(filename, 'r') as fd:
        parameters = yaml.safe_load(fd)

    return parameters
```

So for example when reading the `nav2_params.yaml` file with the `read_yaml` function this is the dictionary that I get (I'm excluding the rest of the other nodes as they have no relevance):

```python
{'controller_server': {'ros__parameters': {'use_sim_time': True,
   'controller_frequency': 20.0,
   'min_x_velocity_threshold': 0.001,
   'min_y_velocity_threshold': 0.5,
   'min_theta_velocity_threshold': 0.001,
   'failure_tolerance': 0.3,
   'progress_checker_plugin': 'progress_checker',
   'goal_checker_plugins': ['general_goal_checker'],
   'controller_plugins': ['FollowPath'],
   'progress_checker': {'plugin': 'nav2_controller::SimpleProgressChecker',
    'required_movement_radius': 0.5,
    'movement_time_allowance': 10.0},
   'general_goal_checker': {'stateful': True,
    'plugin': 'nav2_controller::SimpleGoalChecker',
    'xy_goal_tolerance': 0.25,
    'yaw_goal_tolerance': 0.25},
   'FollowPath': {'plugin': 'dwb_core::DWBLocalPlanner',
    'debug_trajectory_details': True,
    'min_vel_x': 0.0,
    'min_vel_y': 0.0,
    'max_vel_x': 0.26,
    'max_vel_y': 0.0,
    'max_vel_theta': 1.0,
    'min_speed_xy': 0.0,
    'max_speed_xy': 0.26,
    'min_speed_theta': 0.0,
    'acc_lim_x': 2.5,
    'acc_lim_y': 0.0,
    'acc_lim_theta': 3.2,
    'decel_lim_x': -2.5,
    'decel_lim_y': 0.0,
    'decel_lim_theta': -3.2,
    'vx_samples': 20,
    'vy_samples': 5,
    'vtheta_samples': 20,
    'sim_time': 1.7,
    'linear_granularity': 0.05,
    'angular_granularity': 0.025,
    'transform_tolerance': 0.2,
    'xy_goal_tolerance': 0.25,
    'trans_stopped_velocity': 0.25,
    'short_circuit_trajectory_evaluation': True,
    'stateful': True,
    'critics': ['RotateToGoal',
     'Oscillation',
     'BaseObstacle',
     'GoalAlign',
     'PathAlign',
     'PathDist',
     'GoalDist'],
    'BaseObstacle.scale': 0.02,
    'PathAlign.scale': 32.0,
    'PathAlign.forward_point_distance': 0.1,
    'GoalAlign.scale': 24.0,
    'GoalAlign.forward_point_distance': 0.1,
    'PathDist.scale': 32.0,
    'GoalDist.scale': 24.0,
    'RotateToGoal.scale': 32.0,
    'RotateToGoal.slowing_factor': 5.0,
    'RotateToGoal.lookahead_time': -1.0}}},
 'local_costmap': {'local_costmap': {'ros__parameters': {'update_frequency': 5.0,
    'publish_frequency': 2.0,
    'global_frame': 'odom',
    'robot_base_frame': 'base_link',
    'use_sim_time': True,
    'rolling_window': True,
    'width': 3,
    'height': 3,
    'resolution': 0.05,
    'robot_radius': 0.22,
    'plugins': ['voxel_layer', 'inflation_layer'],
    'inflation_layer': {'plugin': 'nav2_costmap_2d::InflationLayer',
     'cost_scaling_factor': 3.0,
     'inflation_radius': 0.55},
    'voxel_layer': {'plugin': 'nav2_costmap_2d::VoxelLayer',
     'enabled': True,
     'publish_voxel_map': True,
     'origin_z': 0.0,
     'z_resolution': 0.05,
     'z_voxels': 16,
     'max_obstacle_height': 2.0,
     'mark_threshold': 0,
     'observation_sources': 'scan',
     'scan': {'topic': '/scan',
      'max_obstacle_height': 2.0,
      'clearing': True,
      'marking': True,
      'data_type': 'LaserScan',
      'raytrace_max_range': 3.0,
      'raytrace_min_range': 0.0,
      'obstacle_max_range': 2.5,
      'obstacle_min_range': 0.0}},
    'static_layer': {'map_subscribe_transient_local': True},
    'always_send_full_costmap': True}},
  'local_costmap_client': {'ros__parameters': {'use_sim_time': True}},
  'local_costmap_rclcpp_node': {'ros__parameters': {'use_sim_time': True}}}}
```

Now that I have all the controller server parameters and the costmap parameters loaded in a Python dictionary I instantiate the Node as follows:

```python
def get_controller_server(namespace=None):
    remappings=[
        ('/tf', f'/{namespace}/tf'),
        ('/tf_static', f'/{namespace}/tf_static'),
    ]

    # Contains the python dictionary shown above.
    parameters = read_yaml('nav2_params.yaml')
    
    controller_server_node = Node(
        namespace=namespace,
        name='nav2_controller_server',
        package='nav2_controller',
        executable='controller_server',
        output='screen',
        
        # ***************************************************************
        # Here! We are passing all the controller_server parameters.
        # How do I also pass the localcost map parameters to the controller node?
        # ***************************************************************
        parameters=[
            parameters['controller_server']['ros__parameters'],
        ],
        remappings=remappings
    )

    return controller_server_node
```

Note that I'm not passing any filenames at all (I'm not relying on any of the helper functions shown in the `nav2_bringup` example), just the parameters that the controller server is supposed to receive.

See the comment above the `parameters` kwarg in the snippet. That's what I'm after. I've already tried passing the local costmap settings as an extra Python dictionary like this:

```python
controller_server_node = Node(
    namespace=namespace,
    name='nav2_controller_server',
    package='nav2_controller',
    executable='controller_server',
    output='screen',
    
    parameters=[
        parameters['controller_server']['ros__parameters'],
        {'local_costmap': parameters['local_costmap']},  # Adding this extra dictionary with the local costmap setttings has no effect.
    ],        
    remappings=remappings
)
```

I checked that indeed this alternative is not feeding the hosted localcost map settings because when I run `ros2 param dump /NAMESPACE/local_costmap/local_costmap` and inspect the generated YAML it doesn' have any of the modified parameters that I'm passing.

Once again thanks for your time and help,
Lucas.

**Comment by SteveMacenski** [comment]
This is an odd use-case and I don't have the 'here's the right answer', but I can provide some information / options. 

The local costmap node is[ within the controller server](https://github.com/ros-navigation/navigation2/blob/main/nav2_controller/src/controller_server.cpp#L68-L71) (all else is eq for the planner server and global costmap) and spins its own executor to do its own thing. Think of it as manual node composition. It'll look for its parameters using the ROS parameter system just like the controller server or any other node in ROS 2. If you have your yaml file

```
local_costmap:
  local_costmap:
    ros__parameters:
      # my parameters
```

then you're all good to go just feeding those into the controller server node in launch since those will be passed onto the manually composed node as well. 

If you want to pass in the actual parameters for the local costmap rather than a file, you'll have to see how it is that the ROS parameter system handles that for composed nodes, I don't know off hand. I suspect just sending it `parameters['controller_server']['ros__parameters'], parameters['local_costmap']['ros__parameters'],` might just work (plus or minus playing around w.r.t. the double nested `local_costmap`). I don't know why `'local_costmap': parameters['local_costmap']` would work - there's no parameter named 'local_costmap`. 

If that doesn't work, which I think it should work (or something like it looking into manual composition parameter passing), then you could instead create a temporary file ([example in python launch](https://github.com/ros-navigation/navigation2/blob/main/nav2_bringup/launch/tb4_simulation_launch.py#L216)) containing the contents of your `dict` and pass that in instead so that the parameters again look like it was from a file after you make your manual modifications. 

If you are dead set against the temporary file solution, then this is more a problem with ROS launch / parameter system that (A) may need some further research on your part about how to make work or (B) find that its not possible and this is a feature request / bug that should be filed and potentially worked on deeper into ROS 2 core. I'd personally recommend trying the snippet I recommended above and if that or a variant of it doesn't work, then using the temporary file. That's not uncommon for autogenerated files / scripts.

**Comment by lucasl84** [comment]
Hi Steve,

I've partially solved that issue by using Jinja templates. What I've done is this: Instead of having a `.yaml` file now I've a `.templ` which essentially is a YAML with some placeholders for Jinja2. So I basically render all the costmap parameters with it and then leave the proper YAML file in the proper output path so it can be read when launching the project:

```python
 def get_nav2_controller_server(namespace=""):
    package_dir_name = 'MY_PACKAGE'
    package_dir = get_package_share_directory(package_dir_name)
    template = os.path.join(package_dir, 'params', 'nav2_controller_server.templ')
    template_context = {
        'global_frame': 'map',
        'robot_base_frame': 'base_link',
        'scan_topic': 'scan',
        'map_topic': '/map_server',
    }

    yaml_filename = os.path.join(
        os.path.dirname(template), 'nav2_controller_server.yaml'
    )
    write_template(
        read_template(template, template_context),
        yaml_filename,
    )

    options = {
        'name': 'nav2_controller_server',
        'package': 'nav2_controller',
        'executable': 'controller_server',
        'parameters': [yaml_filename],
    }

    return Node(**options)
```

However, now I've traded one error for another as for when I launch the project I get:

```bash
[controller_server-8] [INFO] [1721166956.099550687] [nav2_controller_server]: Created progress_checker : progress_checker of type nav2_controller::SimpleProgressChecker
[controller_server-8] [INFO] [1721166956.107470618] [nav2_controller_server]: Created goal checker : goal_checker of type nav2_controller::SimpleGoalChecker
[controller_server-8] [INFO] [1721166956.114757117] [nav2_controller_server]: Controller Server has goal_checker  goal checkers available.
[controller_server-8] [INFO] [1721166956.119380245] [nav2_controller_server]: Created controller : FollowPath of type dwb_core::DWBLocalPlanner
[controller_server-8] [INFO] [1721166956.122331947] [nav2_controller_server]: Setting transform_tolerance to 0.100000
[controller_server-8] [ERROR] [1721166956.145083541] [nav2_controller_server]: Couldn't load critics! Caught exception: No critics defined for FollowPath
[controller_server-8] [ERROR] [1721166956.160497346] []: Caught exception in callback for transition 10
[controller_server-8] [ERROR] [1721166956.160532345] []: Original error: No critics defined for FollowPath
[controller_server-8] [WARN] [1721166956.160580078] []: Error occurred while doing error handling.
[controller_server-8] [FATAL] [1721166956.160593367] [nav2_controller_server]: Lifecycle node nav2_controller_server does not have error state implemented
[lifecycle_manager-13] [ERROR] [1721166956.161228185] [lifecycle_manager]: Failed to change state for node: /nav2_controller_server
[lifecycle_manager-13] [ERROR] [1721166956.161344992] [lifecycle_manager]: Failed to bring up all requested nodes. Aborting bringup.
```

This is despite the fact that I've taken the `nav2_params.yaml` file from the repo from the Galactic branch as reference and adjusted it accordingly (only added the Jinja placeholders for the local costmap section).

Here's a copy of my template (the `{{ VARIABLE_NAME }}` get replaced by Jinja when rendering the template):

```yaml
controller_server:
  ros__parameters:
    use_sim_time: True
    controller_frequency: 20.0
    min_x_velocity_threshold: 0.001
    min_y_velocity_threshold: 0.5
    min_theta_velocity_threshold: 0.001
    failure_tolerance: 0.3
    progress_checker_plugin: progress_checker

    # Original values as from the nav2_params.yaml example
    goal_checker_plugins: ["general_goal_checker"] # "precise_goal_checker"
    controller_plugins: ["FollowPath"]

    # Progress checker parameters
    progress_checker:
      plugin: "nav2_controller::SimpleProgressChecker"
      required_movement_radius: 0.5
      movement_time_allowance: 10.0
    # Goal checker parameters
    #precise_goal_checker:
    #  plugin: "nav2_controller::SimpleGoalChecker"
    #  xy_goal_tolerance: 0.25
    #  yaw_goal_tolerance: 0.25
    #  stateful: True
    general_goal_checker:
      stateful: True
      plugin: "nav2_controller::SimpleGoalChecker"
      xy_goal_tolerance: 0.25
      yaw_goal_tolerance: 0.25
    # DWB parameters
    FollowPath:
      plugin: "dwb_core::DWBLocalPlanner"
      debug_trajectory_details: True
      min_vel_x: 0.0
      min_vel_y: 0.0
      max_vel_x: 0.26
      max_vel_y: 0.0
      max_vel_theta: 1.0
      min_speed_xy: 0.0
      max_speed_xy: 0.26
      min_speed_theta: 0.0
      # Add high threshold velocity for turtlebot 3 issue.
      # https://github.com/ROBOTIS-GIT/turtlebot3_simulations/issues/75
      acc_lim_x: 2.5
      acc_lim_y: 0.0
      acc_lim_theta: 3.2
      decel_lim_x: -2.5
      decel_lim_y: 0.0
      decel_lim_theta: -3.2
      vx_samples: 20
      vy_samples: 5
      vtheta_samples: 20
      sim_time: 1.7
      linear_granularity: 0.05
      angular_granularity: 0.025
      transform_tolerance: 0.2
      xy_goal_tolerance: 0.25
      trans_stopped_velocity: 0.25
      short_circuit_trajectory_evaluation: True
      stateful: True
      critics: ["RotateToGoal", "Oscillation", "BaseObstacle", "GoalAlign", "PathAlign", "PathDist", "GoalDist"]
      BaseObstacle.scale: 0.02
      PathAlign.scale: 32.0
      PathAlign.forward_point_distance: 0.1
      GoalAlign.scale: 24.0
      GoalAlign.forward_point_distance: 0.1
      PathDist.scale: 32.0
      GoalDist.scale: 24.0
      RotateToGoal.scale: 32.0
      RotateToGoal.slowing_factor: 5.0
      RotateToGoal.lookahead_time: -1.0

#controller_server_rclcpp_node:
#  ros__parameters:
#    use_sim_time: True

local_costmap:
  local_costmap:
    ros__parameters:
      update_frequency: 5.0
      publish_frequency: 2.0
      global_frame: {{global_frame}}
      robot_base_frame: {{robot_base_frame}}
      map_topic: {{map_topic}}
      use_sim_time: True
      rolling_window: true
      width: 100
      height: 100
      resolution: 0.04
      robot_radius: 0.22

      # Original values as from the nav2_params.yaml example
      #plugins: ["voxel_layer", "inflation_layer"]
      plugins:
        - static_layer
        - obstacle_layer
        - inflation_layer
      inflation_layer:
        plugin: "nav2_costmap_2d::InflationLayer"
        cost_scaling_factor: 3.0
        inflation_radius: 0.55
      obstacle_layer:
        combination_method: 1
        enabled: true
        footprint_clearing_enabled: true
        max_obstacle_height: 2.0
        observation_sources: ''
        plugin: nav2_costmap_2d::ObstacleLayer
      voxel_layer:
        plugin: "nav2_costmap_2d::VoxelLayer"
        enabled: True
        publish_voxel_map: True
        origin_z: 0.0
        z_resolution: 0.05
        z_voxels: 16
        max_obstacle_height: 2.0
        mark_threshold: 0
        observation_sources: scan
        scan:
          topic: {{scan_topic}}
          max_obstacle_height: 2.0
          clearing: True
          marking: True
          data_type: "LaserScan"
          raytrace_max_range: 3.0
          raytrace_min_range: 0.0
          obstacle_max_range: 2.5
          obstacle_min_range: 0.0
      static_layer:
        map_subscribe_transient_local: True
      always_send_full_costmap: True

  #local_costmap_client:
  #  ros__parameters:
  #    use_sim_time: True

  #local_costmap_rclcpp_node:
  #  ros__parameters:
  #    use_sim_time: True
```

Any insights about why is complaining for the critics even those being there? The commented out sections seem to have no effect as for the same error pop ups if those sections are present or not.

Thanks for your time,
Lucas.

**Comment by SteveMacenski** [comment]
That's a generic error telling me that the parameters aren't being sent correctly. There are no default DWB critics, so if none are specified, then you'd see that error. I see that sometimes when I mistakenly forget to provide a yaml file at all. I think that points to your solution having a bug where none of your parameters are being read in now (or perhaps only the ones in your template?)

**Comment by lucasl84** [comment]
Hi Steve,

Ok... I've just by spotted the issue by pure luck. I guess that resting a bit did its job, to take a fresh look at it. The problem for that error is this (actually with the information that I've provided can be spotted too, BUT is extremely subtle): The `name` parameter for the snippet I've provided above is: `nav2_controller_server` but the key in the YAML file is: `controller_server`. Changing the `name` parameter to `controller_server` in the node creation makes correctly load the configuration present from the YAML file at the `controller_server` key. My damned fault. I'm adding a comment to my lanuch files & YAML files to avoid this type of issue.

Thanks once again for your time Steve!
